modules/dataquery/jsx/definefilters.addfiltermodal.tsx

```diff
@@ -0,0 +1,437 @@
+               crit.visits = selectedVisits || [];
+           }
+
+           props.addQueryGroupItem(
+               props.query,
+               crit,
+           );
+
+           props.closeModal();
+           resolve(null);
+        }
+    );
+    return <Modal title="Add criteria"
+       show={true}
+       throwWarning={true}
+       onClose={props.closeModal}
+       onSubmit={submitPromise}>
+            <div style={{width: '100%', padding: '1em'}}>
+                <h3>Field</h3>
+                <div style={{display: 'flex', width: '100%'}}>
+                    <div style={{width: '40%'}}>
+                    <FilterableSelectGroup groups={props.categories.categories}
+                        mapGroupName={(key) => props.categories.modules[key]}
+                        onChange={props.onCategoryChange} />
+                    </div>
+                    <div style={{width: '100%'}}>
+                    {fieldSelect}
+                    </div>
+                </div>
+                {cardinalityWarning}
+                {criteriaSelect}
+                {visitSelect}
+            </div>
+    </Modal>;
+}
+
+
+/**
+ * Get a list of possible query operators based on a field's dictionary
+ *
+ * @param {object} dict - the field dictionary
+ * @returns {object} - list of options for this dictionary
+ */
+function getOperatorOptions(dict: FieldDictionary) {
+    let options: {[operator: string]: string};
+
+    if (dict.type == 'integer' || dict.type == 'date' ||
+            dict.type == 'interval' || dict.type == 'time' ||
+            dict.type == 'decimal') {
+        // Comparable data types
+        options = {
+            'lt': '<',
+            'lte': '≤',
+            'eq': '=',
+            'neq': '≠',
+            'gte': '≥',
+            'gt': '>',
+        };
+    } else if (dict.type == 'enumeration') {
+        // Enumerations are a dropdown. Comparable operators
+        // are meaningless, but the options are a dropdown
+        // and we might be looking for an option "in" any
+        // of the selected choices.
+        options = {
+            'eq': '=',
+            'neq': '≠',
+            'in': 'in',
+        };
+    } else if (dict.type == 'string' ||
+            dict.type == 'URI') {
+        // We might be looking for a substring.
+        options = {
+            'eq': '=',
+            'neq': '≠',
+            'startsWith': 'starts with',
+            'contains': 'contains',
+            'endsWith': 'ends with',
+        };
+    } else {
+        // fall back to == and !=, valid for any type.
+        options = {'eq': '=', 'neq': '≠'};
+    }
+
+    // Possible cardinalities are unique, single,
+    // optional, or many. Unique and single don't
+    // change the possible operators, optional or
+    // 1-many cardinalities have a couple more
+    // things you can check.
+    if (dict.cardinality == 'optional') {
+        options['isnotnull'] = 'has data';
+        options['isnull'] = ' has no data';
```

Comment:
There is a space before 'has' here but not in any of the other rows. Is this intentional?